### PR TITLE
Make `BatchedSend` restartable

### LINKED ISSUE #5481: Properly support restarting `BatchedSend`
In https://github.com/dask/distributed/issues/5480 we found that when `Worker.batched_stream` is restarted after a broken connection to the scheduler, it enters a broken state where `send` succeeds, but doesn't actually send data (it just sits in the buffer forever).

This is probably fixed by https://github.com/dask/distributed/pull/5457, but I think it may deserve a more thorough fix. If we want `BatchedSend` to be restartable, it should have a clear interface and tests for this.

xref https://github.com/dask/distributed/issues/4133 https://github.com/dask/distributed/issues/4163 https://github.com/dask/distributed/issues/5377

### PR DESCRIPTION
**Pull request by gjoseph92**
This is https://github.com/dask/distributed/pull/6272, with my PR comments addressed.

Closes #5481, #6228, #5480

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -4550,6 +4567,10 @@ async def add_client(
         We listen to all future messages from this Comm.
         """
         assert client is not None
+        assert client not in self.clients, f"Client {client!r} already exists"
```
Comment (gjoseph92) [inline distributed/scheduler.py:4570]: The changes from fb8f1c900874fd4dd39ac7b433d8678d868b86c5 are a driveby (@graingert and I discussed how the current logic is messed up in potentially allowing the wrong client batchedsend to be deleted and closed). I'm happy to pull them out to a separate PR.

**Comment by gjoseph92** [comment]
This is currently plagued by a few reliably failing tests:
```
distributed/tests/test_client.py::test_retire_workers
distributed/tests/test_scheduler.py::test_retire_workers_close
distributed/tests/test_client.py::test_reconnect_timeout
distributed/tests/test_tls_functional.py::test_retire_workers
distributed/tests/test_stress.py::test_stress_scatter_death
```

I can't reproduce these failures locally. Additionally, when I restrict CI to only run these tests, they pass: https://github.com/dask/distributed/runs/6428523276?check_suite_focus=true

Interestingly, `test_retire_workers` seems to be leaking file descriptors:
```
================================ RESOURCE LEAKS ================================
distributed/tests/test_client.py::test_retire_workers leaking fds: leaked 3 file descriptor(s) (16->19)
distributed/tests/test_tls_functional.py::test_retire_workers leaking fds: leaked 1 file descriptor(s) (19->20)
```

**Comment by gjoseph92** [comment]
Update: the `retire_workers` tests do fail on Windows, even when run in isolation. `test_reconnect_timeout` (which is the one that fails most reliably normally) is still not failing in isolation.

**Comment by gjoseph92** [comment]
Thanks to 40e37facf943284a9b09b90e807816d793a59e88, we can now at least see that the asyncio.Task being leaked from `test_reconnect_timeout` is coming from the Client: https://github.com/dask/distributed/runs/6429598771?check_suite_focus=true#step:11:1324

**Comment by fjetter** [comment]
This should be no longer relevant after https://github.com/dask/distributed/pull/6361
